File: src/apps/daemon/internal/runtime/safety/gateway_health.go
package safety

import (
	"sync"
)

type GatewayState int

const (
	GatewayOnline GatewayState = iota
	GatewayDegraded
	GatewayOffline
)

type GatewayMetric struct {
	GatewayIP       string       `json:"gateway_ip"`
	LatencyMs       float64      `json:"latency_ms"`
	PacketLossRatio float64      `json:"packet_loss_ratio"`
	State           GatewayState `json:"state"`
}

type GatewayHealthMonitor struct {
	mu            sync.RWMutex
	gateways      map[string]*GatewayMetric
	lossDegraded  float64
	lossOffline   float64
}

func NewGatewayHealthMonitor() *GatewayHealthMonitor {
	return &GatewayHealthMonitor{
		gateways:     make(map[string]*GatewayMetric),
		lossDegraded: 0.20,
		lossOffline:  0.50,
	}
}

func (m *GatewayHealthMonitor) RegisterGateway(ip string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.gateways[ip] = &GatewayMetric{
		GatewayIP:       ip,
		LatencyMs:       0.0,
		PacketLossRatio: 0.0,
		State:           GatewayOnline,
	}
}

func (m *GatewayHealthMonitor) RecordProbe(ip string, latencyMs, loss float64) {
	m.mu.Lock()
	defer m.mu.Unlock()

	if gw, exists := m.gateways[ip]; exists {
		gw.LatencyMs = latencyMs
		gw.PacketLossRatio = loss
		if loss >= m.lossOffline {
			gw.State = GatewayOffline
		} else if loss >= m.lossDegraded {
			gw.State = GatewayDegraded
		} else {
			gw.State = GatewayOnline
		}
	}
}

func (m *GatewayHealthMonitor) SelectActiveGateway(primary, backup string) string {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if gw, exists := m.gateways[primary]; exists && gw.State == GatewayOnline {
		return primary
	}
	if gw, exists := m.gateways[backup]; exists && gw.State != GatewayOffline {
		return backup
	}
	return primary
}

File: src/packages/control-ui/src/api/gateway_health.ts
export enum GatewayStatus {
  Online = 'online',
  Degraded = 'degraded',
  Offline = 'offline'
}

export interface GatewayMetric {
  ip: string;
  latencyMs: number;
  packetLoss: number;
  status: GatewayStatus;
}

export class GatewayHealthMonitor {
  private readonly gateways = new Map<string, GatewayMetric>();

  constructor(
    private readonly lossDegraded = 0.20,
    private readonly lossOffline = 0.50
  ) {}

  public registerGateway(ip: string): void {
    this.gateways.set(ip, { ip, latencyMs: 0, packetLoss: 0, status: GatewayStatus.Online });
  }

  public recordProbe(ip: string, latencyMs: number, loss: number): void {
    const gw = this.gateways.get(ip);
    if (!gw) return;
    gw.latencyMs = latencyMs;
    gw.packetLoss = loss;
    if (loss >= this.lossOffline) {
      gw.status = GatewayStatus.Offline;
    } else if (loss >= this.lossDegraded) {
      gw.status = GatewayStatus.Degraded;
    } else {
      gw.status = GatewayStatus.Online;
    }
  }

  public selectActiveGateway(primaryIp: string, backupIp: string): string {
    const primary = this.gateways.get(primaryIp);
    if (primary && primary.status === GatewayStatus.Online) {
      return primaryIp;
    }
    const backup = this.gateways.get(backupIp);
    if (backup && backup.status !== GatewayStatus.Offline) {
      return backupIp;
    }
    return primaryIp;
  }
}
